Changes in module 'nf-core/multiqc'
--- modules/nf-core/multiqc/main.nf
+++ modules/nf-core/multiqc/main.nf
@@ -1,16 +1,27 @@
 process MULTIQC {
     label 'process_single'
 
-    conda "${moduleDir}/environment.yml"
+    conda "bioconda::multiqc=1.14"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/multiqc:1.21--pyhdfd78af_0' :
-        'biocontainers/multiqc:1.21--pyhdfd78af_0' }"
+        'https://depot.galaxyproject.org/singularity/multiqc:1.14--pyhdfd78af_0' :
+        'biocontainers/multiqc:1.14--pyhdfd78af_0' }"
 
     input:
-    path  multiqc_files, stageAs: "?/*"
-    path(multiqc_config)
-    path(extra_multiqc_config)
-    path(multiqc_logo)
+    path multiqc_config
+    path extra_multiqc_config
+    path software_versions
+    path workflow_summary
+    path methods_description
+    path multiqc_logo
+    path ('fastqc/raw/*')
+    path ('trimgalore/trim/*')
+    path ('fastqc/trim/*')
+    path ('trimgalore/hardtrim/*')
+    path ('fastqc/hardtrim/*')
+    path ('star/*')
+    path ('picard_alignment_metrics/*')
+    path ('busco/genome/*')
+    path ('busco/annotation/*')
 
     output:
     path "*multiqc_report.html", emit: report

************************************************************
